study1 + assignment1

kkoon9 opened the pull request.

kkoon9 committed:
feat: 2주차 스터디 study 요구사항 [1], [2]

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: StringTddTest 요구사항 [3]

- charAt + exception 처리 테스트 코드 작성

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: CollectionTest 요구사항 [1]

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: 요구사항 [2] - ParameterizedTest

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: CollectionTdd 클래스를 사용하게 코드 변경

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: assignment1 요구사항 README

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: 빈 값을 전달하면 0을 리턴한다 테스트 개발

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: 쉼표(,) 또는 콜론(:)을 구분자로 분리된 배열을 리턴한다 테스트 코드 작성

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: 커스텀 구분자 테스트+ 구분자, 숫자 이외의 값을 전달하는 경우 RuntimeException 발생 테스트

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
feat: 배열의 원소의 합을 리턴하는 테스트 코드 개발

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 commented on build.gradle:
```diff
@@ -11,6 +11,7 @@ repositories {
 
 dependencies {
     testImplementation("org.assertj:assertj-core:3.18.1")
+    testImplementation "org.junit.jupiter:junit-jupiter-params:5.4.2"
```
ParameterizedTest 사용을 위한 의존성 추가 코드입니다.

bradley4 replied:
제가 다 테스트해봤어야했는데 훌륭합니다 .😀

kkoon9 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,6 @@
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.1-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.9-all.zip
```
제 컴퓨터에서 7.1은 계속 build 오류나서 6.9로 낮춰서 진행했습니다.

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.stream.Stream;
+
+public class Calculator {
+    int plus(String input) {
```
- 접근 지시자를 붙여보면 어떨까요?
- 이번에 구현해야하는 것은 문자계산기입니다.  다른 개발자가 유지보수를 맡기로 했을때 
   파라미터로 받는 `input`이라는 변수를 보고 어떤 값일지 예상을 할 수 있을까요?
   `textNumbers`같은 표현은 어떨까요? 동의하시나요?

kkoon9 replied:
`input`보다 직관적인 `textNumbers`으로 변수명을 바꾸는게 좋아보여요!

대니스 접근 지시자에 궁금한 점이 있어서 코멘트 남깁니다!
저는 package-private하게 쓰려고 접근 지시자를 안붙였는데 대니스는 어떤 접근 지시자를 쓰는게 괜찮다고 생각하시나요!?
저는 테스트와 같은 패키지명을 사용하면 public보다는 protected나 default(package-private) 접근 지시자를 사용해도 괜찮다는 생각입니다!
추가로, Calculator 클래스와 CalculatorTest 클래스도 접근 지시자를 public으로 구성하지 않아도 되겠다는 생각을 했습니다!

bradley4 replied:
저는 클래스 레벨에서는 특별한 의도(외부접근을 막지 않으면 안된다는지)가 없지않는 이상 `public`으로 두고 있습니다.
다만 테스트 패키지는 `public`을 Junit5에서 사용하지 않아도 되서 `default`로 두고 있습니다.
일반적으로 개발을 하게되면 보통 public으로 두고 있습니다. default는 패키지 접근을 해야하는 케이스나 코테정도?

클래스 내부 레벨인데요. 여기서부터는 내가 알려줄 정보와 알려주지 않을 정보와 생각해봐야합니다. 내가 외부에 공개하고 싶은 정보는 계산기의 덧셈 기능 뿐입니다. `plus`외 `split`같은 메서드는 외부에서 궁금하지도 알 필요도 없습니다. 캡슐화가 필요할 것같습니다. 따라서 `plus`외 다른 메서드들은 한 클래스에 두어야한다면 `private`으로 두면 어떨까요?

추가적으로 생각해보면 좋은 것은 계산기에서 `split`을 가지고 있어야할까요?  
보통 업무개발을 할때 텍스트 처리 기능을 별도의 클래스에서 가져와서 사용하는데 그러한 점에서 
별도의 클래스를 만들어보면 어떨까요?

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+    int plus(String input) {
+        if (input.isEmpty()) {
+            return 0;
+        }
```
- `null`값이 들어오면 어떻게 되나요?
- 공백이 포함된 문자면 어떻게 해야할까요? `"    "`

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+        if (input.isEmpty()) {
+            return 0;
+        }
+        int[] elements = split(input);
```
- `element`라는 것을 보고 어떤 값인지 예상하기 힘들 것같습니다!

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+                .sum();
+    }
+
+    int[] split(String input) {
```
기능을 나누는 시도가 좋습니다 👍
계산기의 책임은 `계산`을 하는 것이기때문에 이런 문자 분리 기능은 다른 클래스와 협력하는 것으로 바꿔보면 어떨까요?

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    int[] split(String input) {
+        String regex = ",|:";
```
String을 패턴 매칭을 위해 Pattern으로 바꾸는 것은 꽤나 비싼 행위입니다. 미리 Pattern으로 만들어 사용하는 것은 어떨까요? 아래의 글을 참고하세요.
https://stackoverflow.com/questions/2469244/whats-the-difference-between-string-matches-and-matcher-matches/2469275

bradley4 commented on src/test/java/calculator/README.md:
```diff
@@ -0,0 +1,16 @@
+
+- [ ] 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환 (예: “” => 0, "1,2" => 3, "1,2,3" => 6, “1,2:3” => 6)
+- [ ] 앞의 기본 구분자(쉼표, 콜론)외에 커스텀 구분자를 지정할 수 있다. 커스텀 구분자는 문자열 앞부분의 “//”와 “\n” 사이에 위치하는 문자를 커스텀 구분자로 사용한다. 예를 들어 “//;\n1;2;3”과 같이 값을 입력할 경우 커스텀 구분자는 세미콜론(;)이며, 결과 값은 6이 반환되어야 한다.
+- [ ] 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw한다.
```
-[X] 으로 표시하면 알아보기 쉬울 것같아요

bradley4 commented on src/test/java/study/StringTddTest.java:
```diff
@@ -0,0 +1,95 @@
+        char result = stringTdd.charAt(input, index);
+
+        // then
+        Assertions
```
이런건 static import를 해주면 가독성이 더 좋아져요

bradley4 commented on src/test/java/study/StringTddTest.java:
```diff
@@ -0,0 +1,95 @@
+        int[] result = stringTdd.split(input);
+
+        // then
+        Assertions
+                .assertThat(result)
+                .containsExactly(1, 2);
```
디미터 법칙을 제가 말씀드리느랴 예외상황도 설명드려야했는데 제가 못했습니다.
이러한 assertion 같은 경우는 한줄로 쓰는게 가독성이 더 좋습니다.
``` java
assertThat(result).containExactly(1,2);
```

bradley4 commented on src/test/java/study/StringTddTest.java:
```diff
@@ -0,0 +1,95 @@
+        // when
+
+        // then
+        Assertions.assertThatThrownBy(
+                () -> {
+                    stringTdd.charAt(input, index);
+                }
+        )
+                .isInstanceOf(StringIndexOutOfBoundsException.class)
+                .hasMessageContaining("index [%d]는 벗어난 인덱스입니다.", index);
```
👍

bradley4 replied:
``` java
assertThatThrownBy(() -> stringTdd.charAt(input, index))
                .isInstanceOf(StringIndexOutOfBoundsException.class);
```
이런식으로 써보면 가독성이 더 좋아지지않을까요?

bradley4 commented on src/test/java/study/CollectionTddTest.java:
```diff
@@ -0,0 +1,55 @@
+    @Test
+    void split_test_0() {
+        // given
+        CollectionTdd collectionTdd = new CollectionTdd();
+
+        // when
+        int result = collectionTdd.size(numbers);
+
+        // then
+        Assertions
+                .assertThat(result)
+                .isEqualTo(3);
```
실습할때 설명드릴려고 했는데, study는 Junit연습을 위한 것이라 굳이 일급컬렉션을 생각하고 만들지 않으셔도 됩니다~!
``` java
    assertThat(numbers.size()).isEqualTo(3);
```
이정도만 해도 충분합니다

bradley4 commented on src/test/java/study/CollectionTddTest.java:
```diff
@@ -0,0 +1,55 @@
+
+        // then
+        assertTrue(result);
+    }
```
목록으로 입력받은 값의 포함여부 리턴에 대한 테스트도 있지않았나요?
csv source를 써보는 연습이 있었던것같은데 👀

bradley4 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,98 @@
+        boolean result = calculator.isCustomRegex(input);
+
+        // then
+        Assertions.assertThat(result).isEqualTo(true);
```
`isTrue()`를 써도 좋을 것같아요~

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,37 @@
+        input = cleansingBrackets(input);
+
+        String[] inputSplitComma = input.split(",");
+        int[] result =
+                Stream.of(inputSplitComma)
+                        .mapToInt(Integer::parseInt)
+                        .toArray();
```
- 불필요하게 변수를 만들지 않고 리턴하면 더 좋을 것같습니다!
- 파싱하는 원소가 숫자라는 보장이 있을까요?
- 특별한 이유가 있지않는 이상 `List`를 사용하는게 더 좋을 것같습니다.
> [이텍티브자바-배열보다-리스트를사용하라](https://jyami.tistory.com/92)

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,37 @@
+    int[] split(String input) {
+        input = cleansingBrackets(input);
+
+        String[] inputSplitComma = input.split(",");
```
하드코딩을 없애보면 어떨까요? 

see also
> https://tecoble.techcourse.co.kr/post/2020-05-07-avoid-hard-coding/

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,37 @@
+        return input.charAt(index);
+    }
+
+    private String cleansingBrackets(String input) {
```
`cleansing`을 쓰는 것보다 일반적으로 제거의 표현으로 사용하는 `remove`혹은 원의미대로 `subString()`같은 표현을 써보면 다른 개발자들도 이해하기 쉬울 것같습니다.

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,37 @@
+public class StringTdd {
+
+    int[] split(String input) {
+        input = cleansingBrackets(input);
```
해당코드의 input이 안전할 것이라는 보장은 없습니다.

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,37 @@
+        char firstCharacterOfInput = input.charAt(0);
+        char lastCharacterOfInput = input.charAt(input.length() - 1);
+
+        if (firstCharacterOfInput != '(' || lastCharacterOfInput != ')') {
+            return input;
+        }
```
하드코딩을 없애보면 어떨까요?

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+        Pattern pattern = Pattern.compile(regex);
+        Matcher matcher = pattern.matcher(input);
+        return matcher.find();
+
```
불필요한 공백은 제거해주세요

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+            regex = splittingResults[0].replaceAll("//", "");
+        }
+
+        checkException(input, regex);
```
중간중간 변경된 상태를 점검하는 것보다 
변경을하면서 예외를 처리해보면 어떨까요?

이러한 코드는 객체의 상태를 계속 물어보는 코드입니다.

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+
+        checkException(input, regex);
+
+        String[] inputSplitComma = input.split(regex);
```
`inputSplitComma`을 보고 어떤 것을 예측할 수 있을까요?

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,62 @@
+
+        String[] inputSplitComma = input.split(regex);
+        int[] result =
+                Stream.of(inputSplitComma)
+                        .mapToInt(Integer::parseInt)
+                        .toArray();
+        return result;
```
숫자로 안정적으로 변경된다는 보장을 할 수 없습니다.

bradley4 reviewed (changes requested):
아직 과제 2주 남았는데, 이렇게 빨리 하실줄이야 👍
study까지 안하셔도 됐는데 다하셨네요. 제 기억으로는 스터디의 테스트가 하나 빠진것같아요~!

고생하셨습니다. 몇가지 생각해볼만한 코멘트를 남겼습니다.

bradley4 commented:
커밋로그를 확인해봤는데 기능과 커밋이 일치해서 읽기 좋았어요 💯
기왕 연습이니 PR메시지도 연습을 해보면 어떨까요?

bradley4 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,98 @@
+        Assertions.assertThat(result).isEqualTo(0);
+    }
+
+    @DisplayName("쉼표(,) 또는 콜론(:)을 구분자로 분리된 배열을 리턴한다")
+    @Test
+    void split_comma_colon_test() {
+        // given
+        Calculator calculator = new Calculator();
+        String input = "1,2";
```
콜론 테스트는 어디에?

bradley4 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,98 @@
+        Assertions.assertThat(result).containsExactly(1,2);
+    }
+
+    @DisplayName("구분자, 숫자 이외의 값(음수 포함)을 전달하는 경우 RuntimeException 발생하는지에 대한 테스트")
+    @Test
+    void exception_input_test() {
+        // given
+        Calculator calculator = new Calculator();
+        String input = "//;\n1,2";
```
음수에 대한 테스트는 어디에? 👀

bradley4 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,98 @@
+public class CalculatorTest {
+
+    @DisplayName("빈 값을 전달하면 0을 리턴한다")
+    @Test
+    void is_blank_test() {
+        // given
+        Calculator calculator = new Calculator();
+        String input = "";
```
`@NullAndEmptySource` 을 찾아보고 적용해보면 어떨까요?

kkoon9 committed:
fix: 코드리뷰 반영 - 변수명 input보다는 좀 더 직관적인 textNumbers로 변경

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [1]

- null 값을 전달하면 NullPointerException을 던지는 테스트 코드
- 띄어쓰기를 포함한 빈 값을 전달하면 0을 리턴하는 테스트 코드
- 콜론 테스트하는 코드를 위해 ParameterizedTest로 변경
- 음수 테스트하는 코드를 위해 ParameterizedTest로 변경

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [2] - 디미터 법칙 예외 상황

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [3]
- 디미터 법칙 예외 상황 추가 코드

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [4] - isTrue()로 변경

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [5]

- 하드코딩 제거

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [6]

- 직관적인 네이밍 cleansing > remove

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [7]

- 불필요한 공백 제거

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [7]

- 배열보다는 리스트를 사용하게 변경
- 계산기 클래스에 Text 관련 기능 새로운 클래스 TextUtil로 이관

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [8]

- 띄어쓰기일 경우 trim을 이용하여 0으로 출력

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

bradley4 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -20,12 +22,32 @@ void is_blank_test() {
     }
 
-    @DisplayName("쉼표(,) 또는 콜론(:)을 구분자로 분리된 배열을 리턴한다")
+    @DisplayName("null 값을 전달하면 NullPointerException을 발생시킨다")
```
NIT: exception은 보통 던진다고 합니다.

bradley4 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,41 @@
+class CalculatorTest {
+
+    @DisplayName("빈 값을 전달하면 0을 리턴한다")
+    @ParameterizedTest
+    @ValueSource(strings = {"", " "})
```
Junit 5.4 이상에서는 `@EmptySource` 라는 어노테이션이 있습니다.

bradley4 commented on src/test/java/calculator/TextUtilTest.java:
```diff
@@ -0,0 +1,82 @@
+
+    @DisplayName("쉼표(,)와 콜론(:)를 구분자로 분리된 배열을 리턴한다")
+    @ParameterizedTest
+    @ValueSource(strings = {"1,2", "1:2"})
+    void split_comma_colon_test(String input) {
```
👍

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,40 @@
+            return input;
+        }
+
+        return input.substring(1, input.length() - 1);
```
사실 이런 부분들은 `매직넘버`라고 합니다.
`1, length() -1`
1이 의미하는 바가 무엇이고 length() -1 가 무엇인지 맥락을 모르면 알 수 없죠.

> https://javabom.tistory.com/28

bradley4 commented on src/main/java/study/StringTdd.java:
```diff
@@ -0,0 +1,40 @@
+    char charAt(String input, int index) throws StringIndexOutOfBoundsException {
+        if(input.length() <= index) {
+            throw new StringIndexOutOfBoundsException(
+                    String.format("index [%d]는 벗어난 인덱스입니다.", index)
```
`String.format`을 쓴게 좋네요

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -0,0 +1,58 @@
+
+public class TextUtil {
+    static List<Integer> split(String textNumbers) throws NullPointerException {
+        if(textNumbers == null) {
+            throw new NullPointerException("Null을 전달해서는 안 됩니다.");
+        }
```
상황에 맞는 예외를 던지는게 좋습니다.  이런 경우는 인자가 잘못들어온 케이스가 더 맞을 것 같아요
``` suggestion 
       if(textNumbers == null) {
            throw new illegalArgumentException("Null을 전달해서는 안 됩니다.");
        }
```
> [추상화 수준에 맞는 예외를 던지라](https://velog.io/@eversong/Effective-Java-73.-%EC%B6%94%EC%83%81%ED%99%94-%EC%88%98%EC%A4%80%EC%97%90-%EB%A7%9E%EB%8A%94-%EC%98%88%EC%99%B8%EB%A5%BC-%EB%8D%98%EC%A7%80%EB%9D%BC)
> [IllgalArgumentException 문서](http://cris.joongbu.ac.kr/course/java/api/java/lang/IllegalArgumentException.html)

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class TextUtil {
```
텍스트 유틸로 분리해서 책임을 나누셨군요 💯

유틸 클래스의 경우 인스턴스화를 막기 위해 `private` 생성자를 쓰곤 합니다.
> [인스턴스화를 막으려거든 private 생성자를 사용하라](https://madplay.github.io/post/enforce-noninstantiability-with-private-constructor)

kkoon9 replied:
private일 경우 Test 코드를 만들 수 없게 되는데 이같은 경우에는 어떻게 해야할까요?

bradley4 replied:
@kkoon9 
다시 코멘트를 읽어봤는데요. private 생성자를 말씀하시는 걸까요? 
util 클래스의 메서드는 static으로 사용되서 테스트하는데는 문제가 없을 것같아요!

private method 테스트하는 것을 질문하신거면 다음 모임때 상세하게 답변드리겠습니다.

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.stream.Collectors;
+
+public class TextUtil {
+    static List<Integer> split(String textNumbers) throws NullPointerException {
```
여기에도 접근 제어자가 있다면 좋을 것 같습니다!

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -0,0 +1,58 @@
+        checkException(textNumbers, regex);
+
+        String[] textNumbersSplitComma = textNumbers.split(regex);
+        List<Integer> result =
+                Arrays.stream(textNumbersSplitComma)
+                        .mapToInt(Integer::parseInt)
+                        .boxed()
+                        .collect(Collectors.toList());
+
+        return result;
```
불필요한 변수는 제거해보면 좋을 것같아요!

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -0,0 +1,58 @@
+            regex = splittingResults[0].replaceAll("//", "");
+        }
+
+        checkException(textNumbers, regex);
```
Tell, don't ask 원칙이라는 글을 참고해보면 좋을 것같아요
> https://effectiveprogramming.tistory.com/entry/Tell-dont-ask

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -0,0 +1,58 @@
+        if (textNumbers.trim().isEmpty()) {
+            return Collections.EMPTY_LIST;
+        }
+        String regex = ",|:";
```
하드코딩을 제거해보면 어떨까요?

bradley4 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,12 @@
+import static calculator.TextUtil.split;
+
+class Calculator {
+    int plus(String textNumbers) {
+        List<Integer> numbers = split(textNumbers);
+        return numbers.stream().mapToInt(num -> num).sum();
+    }
```
여기도 굳이 접근 지시자를 사용하지 않을 이유는 없을 것같아요~
``` suggestion
   public int plus(String textNumbers) {
        List<Integer> numbers = split(textNumbers);
        return numbers.stream()
                .mapToInt(num -> num)
                .sum();
    }

bradley4 reviewed (commented):
요구사항이 많았는데 전반적으로 잘 지켜주셨습니다.
더 수정해보셔도 좋고, 마무리 요청을 하시면 머지하겠습니다.

kkoon9 committed:
fix: 코드리뷰 반영 [9]

- "", " " 을 포함하는 EmptySource 어노테이션이 존재하므로 수정

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드리뷰 반영 [10]

- argument 에러이므로 추상화 수준에 맞는 에러로 변경

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드리뷰 반영 [11] - 불필요한 변수 제거

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드리뷰 반영 [12] - 하드코딩 제거

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드리뷰 반영 [13]

- TDA가 이렇게 하는건지는 잘 모르겠네요

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영[13-2] - TDA

- 리뷰 반영 [13] 과 [13-2] 중 긴가민가하네요

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영[13-2] - TDA [추가 수정]

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

kkoon9 committed:
fix: 코드 리뷰 반영 [14] - 매직넘버 제거

Signed-off-by: kkoon9 <rndrnjs2003@gmail.com>

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -8,9 +8,9 @@
 
 public class TextUtil {
-    static List<Integer> split(String textNumbers) throws NullPointerException {
+    static List<Integer> split(String textNumbers) throws IllegalArgumentException {
```
개인적인 궁금한 점인데, `checked exception` 처럼 `throws`를 명시하는 이유가 있을까요?
취향의 영역일 수도 있지만 코멘트를 남겨둡니다.

bradley4 commented on src/main/java/calculator/TextUtil.java:
```diff
@@ -26,8 +28,16 @@
+
+        if (!matcher.find()) {
+            throw new RuntimeException("구분자, 숫자 이외의 값(음수 포함)을 포함합니다.");
+        }
```
기존 코드를 유지하는 방향에서라면, Pattern 정도를  method로 따로 만들어두고 
`matcher.find()`로 체크할 수 있겠네요.

사실 제가 원했던 것은 원시값을 포장한 `PostiveNumber` 객체를 만들고 객체 생성시점에 체크해서 
나의 진행 상태를 묻지않고 객체 생성을 통해 검증하는 것인데 이론적인 이야기고 산으로 가는 것같네요.  😵

다만, 확인해보시면 좋을 부분은 대체적으로 코드에서 값을 만들고 
다시 값을 꺼내와서 검증하는 행위가 반복되고 있네요.
다음에는 이런 부분을 어떻게 줄일 수 있을지 고민해보면 좋을 것같아요.  이번 연습은 이정도만 해도 괜찮을 것같습니다.

그리고 메서드가 계속 길어지고 있는데요. 연습을 하실때는 제약을 걸어보면서 개발을 해보면 좋습니다.
클래스와 메서드의 크기를 작게 유지해보라는 권고사항이 있었던 걸로 기억하는데요. 
메서드 레벨에서 20라인이라든지 그런식의 제약을 두면서 연습을 해보면 좋습니다.

bradley4 replied:
단, 제가 자세히 가이드를 드리면 학습을 진행해주시는 의미가 반감되고 
구현에 정답은 없으므로 요 방식으로 진행하셔도 될 것 같습니다!

bradley4 commented:
이번 과제를 진행하시면서 개발자로서 성장하는데 조금이나마 도움이 되었기를 바랍니다.
정말 고생하셨습니다 🙇‍♂️
